src/subgraph/service.ts

```
@@ -0,0 +1,822 @@
+        operation,
+        `Query execution failed: ${(error as Error).message}`,
+        { cause: error }
+      )
+    }
+  }
+
+  /**
+   * Converts CommP to hex string
+   */
+  private commPToHex (commP: CommP): string {
+    return toHex(CID.parse(commP.toString()).bytes)
+  }
+
+  /**
+   * Transforms provider data to ApprovedProviderInfo
+   */
+  private transformProviderData (data: any): ApprovedProviderInfo {
+    return {
+      owner: data.address != null && data.address !== '' ? data.address : data.id,
+      pdpUrl: data.pdpUrl,
+      pieceRetrievalUrl: data.pieceRetrievalUrl,
+      registeredAt: this.parseTimestamp(data.registeredAt),
+      approvedAt: this.parseTimestamp(data.approvedAt)
+    }
+  }
+
+  /**
+   * Safely parses timestamp values
+   */
+  private parseTimestamp (value?: number | string): number {
+    if (value == null) return 0
+    const parsed = Number(value)
+    return isNaN(parsed) || parsed === 0 ? 0 : parsed
+  }
+
+  /**
+   * Validates provider data completeness
+   */
+  private isValidProviderData (data: any): boolean {
+    return (
+      data?.id != null &&
+      data.id.trim() !== '' &&
+      data?.pdpUrl != null &&
+      data.pdpUrl.trim() !== '' &&
+      data?.pieceRetrievalUrl != null &&
+      data.pieceRetrievalUrl.trim() !== ''
+    )
+  }
+
+  /**
+   * Queries the subgraph to find approved storage providers that have a specific piece (CommP).
+   *
+   * It sends a GraphQL query to the configured endpoint and parses the response to extract
+   * a list of providers, including their addresses and retrieval URLs.
+   *
+   * @param commp - The piece commitment (CommP) to search for.
+   * @returns A promise that resolves to an array of `ApprovedProviderInfo` objects.
+   *          Returns an empty array if no providers are found or if an error occurs during the fetch.
+   */
+  async getProvidersForCommP (commP: CommP): Promise<ApprovedProviderInfo[]> {
+    const hexCommP = this.commPToHex(commP)
```

it might be better if you do the `asCommP` (and `=== null` check) here that I had above in `commPToHex`, then you can return an error from here if you don't get one, and `commPToHex` can just assume a `CommP` (in fact, you probably don't even need that function since it's just a `toHex(commP.bytes)`

---

I can get rid of `commPToHex` function to simply replace it with `toHex(commP.bytes)`